binary-search.py

```diff
@@ -2023,8 +2023,8 @@ def main():
                              do_search = True
                              next_rate = rate # since this was only the sniff test, keep the current rate
                         else:
-                             prev_pass_rate.append(rate) # add the newly passed rate to the stack of passed rates; this will become the new reference for passed rates
-                             next_rate = (prev_fail_rate + rate) / 2
+                             prev_pass_rate.append(tx_rate) # add the newly passed rate to the stack of passed rates; this will become the new reference for passed rates
```

Comment:
The calculation of tx_rate is done inside a loop that iterates through the device pairs that are used -- so the tx_rate that is used here is only representative of the final device pair that was checked.  Depending on the test and/or DUT the various pairs might be having significant different behavior so I think some more thought needs to be put in to how to coalesce the various pairs.

Also, should the same logic be used in both the pass and fail case?  At the very least I think there should be some comments that explain why the logic is deviating from what is currently there (ie. why tx_rate instead of rate) and also why the pass/fail logic is the same or different.

Reply:
> The calculation of tx_rate is done inside a loop that iterates through the device pairs that are used -- so the tx_rate that is used here is only representative of the final device pair that was checked. Depending on the test and/or DUT the various pairs might be having significant different behavior so I think some more thought needs to be put in to how to coalesce the various pairs.

This is a good point. I see 2 options to address this:
1) Take the minimum achieved rate across the device pairs;
2) Be conservative and take (rate - rate_tolerance)

I'm more inclined to option 1).
Also, I wonder if option 2) could have some corner case issue interacting with search_granularity... probably not, but I didn't check.

Another point that I didn't take into account is the case with rate_unit = %... just taking tx_rate will mix units. I will try to address that as well... maybe converting tx_rate between mpps and %.

> 
> Also, should the same logic be used in both the pass and fail case? At the very least I think there should be some comments that explain why the logic is deviating from what is currently there (ie. why tx_rate instead of rate) and also why the pass/fail logic is the same or different.

We are using tx_rate instead of rate because when a test passes, we assume that
the system is capable of reaching that rate (with an acceptable amount of
packet loss)... but because we take into account rate_tolerance, it could be
that the system CANNOT reach that rate successfully... it can actually reach a value
somewhere between rate - rate_tolerance and rate... and that value is tx_rate.

Here is an example of what could happen for a system that can reach at maximum 46mpps with acceptable packet loss:

rate_tolerance = 10
search_granularity = 6
initial_rate = 100mpps

rate 100mpps: fail
rate 50mpps: pass (with tx_rate = 46mpps, which is within the rate_tolerance)
Notice here that we save 50mpps as a sucessfull rate, when we actually reached only 46mpps
rate 75mpps: fail
rate 62.5mpps: fail
rate 56.25mpps: fail
rate 53.125mmps (final validation): fail

I hope it is clear from the above example that saving rate instead of tx_rate
with a rate_tolerance > 0 can restrict the binary search to an unreachable
interval, causing the search to fail when it shouldn't.

I don't think we should use the same logic (using tx_rate instead of rate) in
the fail case. First, because we don't have to, since rate is already a
conservative choice for the fail case. And second, because there are multiple
variables in play in the case of fail that could have been the reason to fail:
tx_rate, packet loss, runtime, duplicate packets, etc.